Better `INSTALL.md` after Dune switch. (#14677)
coq/coq: INSTALL.md

Installing From Sources
=======================

> The recommended way to install and use Coq is to install the Coq
> platform or to rely on a package manager (e.g. opam or Nix).

Reviewer: ```suggestion
We recommend using a package manager such as opam or Nix to install the Coq platform.
```

Author: This changes the meaning completely, so nack.

Reviewer: Yeah, I missed a bit of the original meaning.  My suggestion was mainly to begin with "We recommend ..."

See https://coq.inria.fr/download and
https://github.com/coq/coq/wiki#coq-installation to learn more.

> Otherwise, the remainder of this file gives technical instructions for
> building Coq from sources.

Reviewer: ```suggestion
Otherwise, the remainder of this file explains how to build Coq from sources.
```

Author: The goal of this introduction was to discourage normal users from reading this file / attempting to install from sources. It was already alleviated by the suggestion of Hugo that I've applied (leading to the current sentence). If you remove "technical" from it, then it loses the word that was supposed to be discouraging. I'm open to other proposals.

Reviewer: Maybe a bit much to expect from the single word "technical".
How about this?  Too unfriendly?  OTOH, I don't want to discourage anyone who is willing to spend the time to learn about Coq internals.

```suggestion
Otherwise, the remainder of this file explains how to build Coq from sources.  We recommend using this approach only if you plan to make changes to Coq and are willing to work through any issues with limited assistance.  
```

Author: That's too much indeed. We are willing to provide as much assistance as people need if they want to contribute. I'll push some changes. Let me know what you think.

Installing From Sources
=======================

> If you are just a user, you should probably install Coq using the
> platform or a package manager (e.g. opam or Nix) instead.

Reviewer 2: To oppose usages rather than classifying people, I would suggest:
```suggestion
The recommended way to install and use Coq is to install the Coq
platform or to rely on a package manager (e.g. opam or Nix).
```

See https://coq.inria.fr/download and
https://github.com/coq/coq/wiki#coq-installation to learn more.

> If you are a developer or you want to package Coq or you really need
> to build Coq from sources without a package manager like opam or Nix,
> then read on.

Reviewer 2: The "technical" below is intended to express the "discourage" that you want to express:
```suggestion
Otherwise, the remainder of this file gives technical instructions for
building Coq from sources.
```

Installing From Sources
=======================

To install and use Coq, we recommend relying on [the Coq
platform](https://github.com/coq/platform/) or on a package manager
(e.g. opam or Nix).

See https://coq.inria.fr/download and
https://github.com/coq/coq/wiki#coq-installation to learn more.

If you really need to build Coq from sources manually (e.g. to
contribute to Coq or to write a Coq package), then the remainder of this

> file explains how to.

Reviewer 3: ```suggestion
file explains how to do this.
```
Sounded incomplete to my ear.

> file explains how to.

Reviewer 3: ```suggestion
file explains how to do this.
```
Sounded incomplete to my ear.

Reviewer: Yup.  Or could just be "... explains how."

> If you really need to build Coq from sources manually (e.g. to
> contribute to Coq or to write a Coq package), then the remainder of this
> file explains how to.

Reviewer: ```suggestion
If you need to build Coq from sources manually (e.g. to
contribute to Coq or to write a Coq package), the remainder of this
file explains how to do so.
```

"really" is not really needed here.  It doesn't really add much.

can be used; see the entry about `-flambda-opts` in the build guide
for more details.

Choice of Build and Installation Procedure
------------------------------------------

There are two partially overlapping infrastructures available to build
Coq. They are available through `Makefile.make` (legacy / hybrid

> build) and `Makefile.dune` (full Dune build).

Reviewer 3: Since `Makefile.make` is the legacy option, perhaps it should be mentioned after `Makefile.dune`?

Author: I mentioned it first because it's the default.

> There are two partially overlapping infrastructures available to build
> Coq. They are available through `Makefile.make` (legacy / hybrid
> build) and `Makefile.dune` (full Dune build).

Reviewer: ```suggestion
There are two partially overlapping ways to build
Coq. They are through `Makefile.make` (legacy / hybrid
build) and `Makefile.dune` (full Dune build).
```

Author: Not convinced by the replacement of "infrastructures" (which I used in place of the previous "build systems" which was ambiguous / created confusion) by the less specific "ways"

Reviewer: IMHO, it's a little unusual, but entirely unknown, to use "infrastructure" in its plural form.  It's an undefined term, so hardly specific. It leads the reader to wonder what's in each "infrastructure", if they even care.  Also unclear what a "hybrid" build is.  And why do readers care that they overlap?

Might be clearer to say Makefile.make is the older build system.  Makefile.dune primarily uses dune, except that it invokes Makefile.make to build a few targets.

Do whatever you think is best.

> You can use the variable `COQ_USE_DUNE` to change which infrastructure

Reviewer 3: What kind of variable? I'm guessing environment.

to use by default. This is useful for Coq development, where we
recommend to rely mainly on `Makefile.dune`.

In both cases, the OCaml parts are built using
[Dune](https://github.com/ocaml/dune). The main difference between the

> two infrastructures is how the `.vo` files are built.

Reviewer: ```suggestion
two ways is how the `.vo` files are built.
```

In the case of `Makefile.make`, `.vo` files are built with a legacy
Makefile, similar to what `coq_makefile` would do. In the case of
`Makefile.dune`, `.vo` files are built with Dune, thanks to its
recently-added, and still experimental, Coq mode.

See the documentation for the two infrastructures:

- [Legacy build](dev/doc/INSTALL.make.md)
- [Full Dune build](dev/doc/build-system.dune.md)

> See also [`dev/doc/README.md`](dev/doc/README.md).

Reviewer 2: Looks very nice!

Optionally:
- add that an practical difference is also that `Makefile.dune` calls configuration itself
- that it is generally better not to mix the two systems [experimentally, at least due to the `_build_vo` directory and to the existence of a `config/coq_config.ml`]

Author: >add that an practical difference is also that Makefile.dune calls configuration itself

`./configure` is not mentioned in this file. It is in both build infrastructure documentations. (BTW, it is possible to manually call `./configure` with `Makefile.dune`, it's just not recommended for dev mode.)

>that it is generally better not to mix the two systems [experimentally, at least due to the _build_vo directory and to the existence of a config/coq_config.ml]

OK, done.

[Dune](https://github.com/ocaml/dune), and a makefile-based one
[similar to `coq_makefile`].

See the documentation for the two build systems at:

- [Legacy build](dev/doc/INSTALL.make.md)
- [Full Dune build](dev/doc/build-system.dune.md)

> For Coq development, it is recommended to rely on the full Dune build.

Reviewer: ```suggestion
For Coq development, we recommend using the full Dune build.
```

[Dune](https://github.com/ocaml/dune), and a makefile-based one
[similar to `coq_makefile`].

See the documentation for the two build systems at:

- [Legacy build](dev/doc/INSTALL.make.md)
- [Full Dune build](dev/doc/build-system.dune.md)

> For Coq developers, it is recommended to rely on the full Dune build.

Reviewer 2: ```suggestion
For Coq development, it is recommended to rely on the full Dune build.
```
